astropy/coordinates/baseframe.py

```diff
@@ -424,39 +424,114 @@ def __init__(self, *args, copy=True, representation_type=None,
         representation_type = kwargs.pop('representation_type', representation_type)
 
         if representation_type is not None or differential_type is not None:
+            representation_type, differential_type = self._infer_representation_differential_type(
+                representation_type, differential_type)
+            self.set_representation_cls(representation_type, **differential_type)
+        else:
+            self._representation = {'base': self.default_representation,
+                                    's': self.default_differential}
```

Comment:
I was not entirely satisfied with this mix between `self.set_representation_cls(...)` and `self._representation = ...`, but this chunk of code:

```python
        if base is None:
            base = self._representation['base']
```

makes `set_representation_cls` a bit obtuse, and I'm still not sure how to untangle it.